File: csharp-selenium-nunit-reqnroll/SauceDemo.UiTests/Pages/LoginPage.cs
using OpenQA.Selenium;

namespace SauceDemo.UiTests.Pages;

public sealed class LoginPage : BasePage
{
    private static readonly By UsernameInput = By.CssSelector("[data-test='username']");
    private static readonly By PasswordInput = By.CssSelector("[data-test='password']");
    private static readonly By LoginButton = By.CssSelector("[data-test='login-button']");
    private static readonly By ErrorBanner = By.CssSelector("[data-test='error']");

    public LoginPage(IWebDriver driver) : base(driver)
    {
    }

    /// <summary>Submits the login form without asserting the outcome - the caller decides whether success or a specific error is expected.</summary>
    public void SubmitLogin(string username, string password)
    {
        Type(UsernameInput, username);
        Type(PasswordInput, password);
        Click(LoginButton);
    }

    public string GetErrorMessage() => TextOf(ErrorBanner);

    public bool HasError() => WaitAndCheckVisible(ErrorBanner);
}

File: ts-playwright-cucumber/src/pages/loginPage.ts
import type { Page } from 'playwright';
import { BasePage } from './basePage.ts';

const USERNAME_INPUT = "[data-test='username']";
const PASSWORD_INPUT = "[data-test='password']";
const LOGIN_BUTTON = "[data-test='login-button']";
const ERROR_BANNER = "[data-test='error']";

export class LoginPage extends BasePage {
  constructor(page: Page) {
    super(page);
  }

  /** Submits the login form without asserting the outcome - the caller decides whether success or a specific error is expected. */
  async submitLogin(username: string, password: string): Promise<void> {
    await this.type(USERNAME_INPUT, username);
    await this.type(PASSWORD_INPUT, password);
    await this.click(LOGIN_BUTTON);
  }

  getErrorMessage(): Promise<string> {
    return this.textOf(ERROR_BANNER);
  }

  hasError(): Promise<boolean> {
    return this.waitAndCheckVisible(ERROR_BANNER);
  }
}
